projects/packages/autoloader/tests/php/lib/class-test-plugin-factory.php

```diff
@@ -0,0 +1,468 @@
+		file_put_contents(
+			$plugin_dir . DIRECTORY_SEPARATOR . 'composer.json',
+			json_encode( $composer_config, JSON_PRETTY_PRINT | JSON_UNESCAPED_SLASHES )
+		);
+
+		// We're assuming composer is installed and hope failures are distinct!
+		exec( 'composer install -d ' . escapeshellarg( $plugin_dir ) . ' 2>&1' );
+		if ( ! is_file( $plugin_dir . DIRECTORY_SEPARATOR . 'vendor' . DIRECTORY_SEPARATOR . 'autoload.php' ) ) {
+			throw new \RuntimeException( 'Unable to execute the `composer` command for tests.' );
+		}
+
+		if ( isset( $this->autoloader_version ) && ! is_file( $plugin_dir . DIRECTORY_SEPARATOR . 'vendor' . DIRECTORY_SEPARATOR . 'autoload_packages.php' ) ) {
+			throw new \RuntimeException( 'Failed to install the autoloader.' );
+		}
+
+		// Local autoloaders require using the branch but we may not want to treat it as a developer build.
+		if ( $this->is_using_local_package() ) {
+			$manifest_dir = $plugin_dir . DIRECTORY_SEPARATOR . 'vendor' . DIRECTORY_SEPARATOR . 'composer' . DIRECTORY_SEPARATOR;
+			$manifests    = array( 'jetpack_autoload_classmap.php', 'jetpack_autoload_psr4.php', 'jetpack_autoload_psr0.php', 'jetpack_autoload_filemap.php' );
+			foreach ( $manifests as $manifest ) {
+				$manifest = $manifest_dir . $manifest;
+				if ( ! is_file( $manifest ) ) {
+					continue;
+				}
+
+				$content = file_get_contents( $manifest );
+				// Use a sufficiently large version so that the local package will always be the latest autoloader.
+				$content = str_replace( 'dev-master', self::VERSION_CURRENT, $content );
+				file_put_contents( $manifest, $content );
+			}
+		}
+
+		return $plugin_dir;
+	}
+
+	/**
+	 * Indicates whether or not we are linking to the local package.
+	 *
+	 * @return bool
+	 */
+	private function is_using_local_package() {
+		return self::CURRENT === $this->autoloader_version;
+	}
+
+	/**
+	 * Creates and returns the configuration for the composer.json file.
+	 *
+	 * @return array
+	 */
+	private function build_composer_config() {
+		$composer_config = array(
+			'name'     => 'testing/' . $this->slug,
+			'autoload' => $this->autoloads,
+		);
+		if ( $this->is_using_local_package() ) {
+			$composer_config['require']      = array( 'automattic/jetpack-autoloader' => 'dev-master' );
+			$composer_config['repositories'] = array(
+				array(
+					'type'    => 'path',
+					'url'     => TEST_PACKAGE_DIR,
+					'options' => array(
+						'symlink' => true,
+					),
+				),
+			);
+		} elseif ( isset( $this->autoloader_version ) ) {
+			$composer_config['require'] = array( 'automattic/jetpack-autoloader' => $this->autoloader_version );
+		}
+
+		if ( isset( $this->composer_options ) ) {
+			$composer_config = array_merge( $composer_config, $this->composer_options );
+		}
+
+		return $composer_config;
+	}
+
+	/**
+	 * Recursively removes a directory and all of its files.
+	 *
+	 * @param string $dir The directory to remove.
+	 */
+	private function remove_directory( $dir ) {
```

Comment:
I don't know if it may be a good idea to have a separate script to clean up the `tmp` folder, maybe even a composer script? Just to make sure we're not leaving any artifacts behind when restarting a failed test.

Reply:
In an earlier revision this actually just called `rm -rf` but in the interest of supporting Windows this alternative was used. I'll highlight that the directory is only ever removed if the checksum changes and so I don't know that we would want something more heavy-handed for cleanup. The idea was to avoid installing a dummy plugin and autoloader if an identical version of that same plugin and autoloader already exists on disk.